pixy/consumer_test.go

```diff
@@ -0,0 +1,486 @@
+package pixy
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/mailgun/kafka-pixy/Godeps/_workspace/src/github.com/mailgun/log"
+	"github.com/mailgun/kafka-pixy/Godeps/_workspace/src/github.com/mailgun/sarama"
+	. "github.com/mailgun/kafka-pixy/Godeps/_workspace/src/gopkg.in/check.v1"
+)
+
+type SmartConsumerSuite struct {
+	producer *GracefulProducer
+}
+
+var _ = Suite(&SmartConsumerSuite{})
+
+func (s *SmartConsumerSuite) SetUpSuite(c *C) {
+	InitTestLog()
+	var err error
+	config := NewConfig()
+	config.ClientID = "producer"
+	config.Kafka.SeedPeers = testKafkaPeers
+	config.ChannelBufferSize = 1
+	s.producer, err = SpawnGracefulProducer(config)
+	c.Assert(err, IsNil)
+}
+
+func (s *SmartConsumerSuite) TearDownSuite(c *C) {
+	s.producer.Stop()
+}
+
+func (s *SmartConsumerSuite) TestResolveAssignments(c *C) {
+	c.Assert(resolveAssignments(nil, nil), IsNil)
+	c.Assert(resolveAssignments(nil, []string{}), IsNil)
+	c.Assert(resolveAssignments(nil, []string{"a"}), IsNil)
+	c.Assert(resolveAssignments(nil, []string{"a", "b"}), IsNil)
+	c.Assert(resolveAssignments([]int32{}, nil), IsNil)
+	c.Assert(resolveAssignments([]int32{}, []string{}), IsNil)
+	c.Assert(resolveAssignments([]int32{}, []string{"a"}), IsNil)
+	c.Assert(resolveAssignments([]int32{}, []string{"a", "b"}), IsNil)
+	c.Assert(resolveAssignments([]int32{1}, nil), IsNil)
+	c.Assert(resolveAssignments([]int32{1}, []string{}), IsNil)
+
+	c.Assert(resolveAssignments([]int32{0}, []string{"a"}),
+		DeepEquals, map[string]map[int32]none{
+			"a": map[int32]none{0: nothing},
```

Comment:
I think `map[T]bool` is a relatively common way of doing sets in Go. It might be worth doing it that way rather than `map[T]none`, since readers may be more likely to recognize it. I think it also looks more intuitive to write:

```go
myMap[someKey] = true
```

instead of

```go
myMap[someKey] = nothing
```